web_poet/rules.py

```diff
@@ -0,0 +1,324 @@
+from __future__ import annotations  # https://www.python.org/dev/peps/pep-0563/
+
+import importlib
+import importlib.util
+import pkgutil
+import warnings
+from collections import deque
+from operator import attrgetter
+from typing import Any, Dict, Iterable, List, Optional, Type, TypeVar, Union
+
+import attrs
+from url_matcher import Patterns
+
+from web_poet._typing import get_item_cls
+from web_poet.pages import ItemPage
+from web_poet.utils import _create_deprecated_class, as_list, str_to_pattern
+
+Strings = Union[str, Iterable[str]]
+
+PageObjectRegistryTV = TypeVar("PageObjectRegistryTV", bound="PageObjectRegistry")
+
+
+@attrs.define(frozen=True)
+class ApplyRule:
+    """A single override rule that specifies when a Page Object should be used
+    in lieu of another.
+
+    This is instantiated when using the :func:`web_poet.handle_urls` decorator.
+    It's also being returned as a ``List[ApplyRule]`` when calling the
+    ``web_poet.default_registry``'s :meth:`~.PageObjectRegistry.get_rules`
+    method.
+
+    You can access any of its attributes:
+
+        * ``for_patterns`` - contains the list of URL patterns associated with
+          this rule. You can read the API documentation of the `url-matcher
+          <https://url-matcher.readthedocs.io/>`_ package for more information
+          about the patterns.
+        * ``use`` - The Page Object that will be **used**.
+        * ``instead_of`` - The Page Object that will be **replaced**.
+        * ``to_return`` - The data container class that the Page Object returns.
+        * ``meta`` - Any other information you may want to store. This doesn't
+          do anything for now but may be useful for future API updates.
+
+    .. tip::
+
+        The :class:`~.ApplyRule` is also hashable. This makes it easy to store
+        unique rules and identify any duplicates.
+    """
+
+    for_patterns: Patterns = attrs.field(converter=str_to_pattern)
+    use: Type[ItemPage]
```

Comment:
What do you think about `use` argument keyword-only?

Reply:
As well as all other arguments, except for `for_patterns`?

Reply:
It won't be backwards compatible (unless we make fix it somehow), but it may be ok, we might just document it in the changelog.

Reply:
That's a good idea! I believe kwargs enforces better readability in PO codebases especially for distinguishing between `use` and `instead_of`. https://github.com/scrapinghub/web-poet/pull/84/commits/9b6f9c4d836b0335850c076e1eed66374d18ecc6